pkg/app/piped/executor/lambda/rollback.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+type rollbackExecutor struct {
+	executor.Input
+}
+
+func (e *rollbackExecutor) Execute(sig executor.StopSignal) model.StageStatus {
+	var (
+		ctx            = sig.Context()
+		originalStatus = e.Stage.Status
+		status         model.StageStatus
+	)
+
+	switch model.Stage(e.Stage.Name) {
+	case model.StageRollback:
+		status = e.ensureRollback(ctx)
+	default:
+		e.LogPersister.Errorf("Unsupported stage %s for cloudrun application", e.Stage.Name)
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	return executor.DetermineStageStatus(sig.Signal(), originalStatus, status)
+}
+
+func (e *rollbackExecutor) ensureRollback(ctx context.Context) model.StageStatus {
+	// Not rollback in case this is the first deployment.
+	if e.Deployment.RunningCommitHash == "" {
+		e.LogPersister.Errorf("Unable to determine the last deployed commit to rollback. It seems this is the first deployment.")
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	runningDS, err := e.RunningDSP.GetReadOnly(ctx, e.LogPersister)
+	if err != nil {
+		e.LogPersister.Errorf("Failed to prepare running deploy source data (%v)", err)
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	deployCfg := runningDS.DeploymentConfig.LambdaDeploymentSpec
+	if deployCfg == nil {
+		e.LogPersister.Errorf("Malformed deployment configuration: missing LambdaDeploymentSpec")
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	cloudProviderName, cloudProviderCfg, found := findCloudProvider(&e.Input)
+	if !found {
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	fm, ok := loadFunctionManifest(&e.Input, deployCfg.Input.FunctionManifestFile, runningDS)
+	if !ok {
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	if !rollback(ctx, &e.Input, cloudProviderName, cloudProviderCfg, fm) {
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	return model.StageStatus_STAGE_SUCCESS
+}
+
+func rollback(ctx context.Context, in *executor.Input, cloudProviderName string, cloudProviderCfg *config.CloudProviderLambdaConfig, fm provider.FunctionManifest) bool {
+	in.LogPersister.Infof("Start rollback the lambda function: %s to original stage", fm.Spec.Name)
+	client, err := provider.DefaultRegistry().Client(cloudProviderName, cloudProviderCfg, in.Logger)
+	if err != nil {
+		in.LogPersister.Errorf("Unable to create Lambda client for the provider %s: %v", cloudProviderName, err)
+		return false
+	}
+
+	// Restore original traffic config from metadata store.
+	originalTrafficKeyName := fmt.Sprintf("%s-%s-original", fm.Spec.Name, in.Deployment.RunningCommitHash)
+	originalTrafficCfgData, ok := in.MetadataStore.Get(originalTrafficKeyName)
+	if !ok {
+		in.LogPersister.Errorf("Unable to prepare original traffic config to rollback Lambda function %s. It seems this is the first deployment.", fm.Spec.Name)
+		return false
+	}
+
+	originalTrafficCfg := provider.RoutingTrafficConfig{}
+	if err := json.Unmarshal([]byte(originalTrafficCfgData), &originalTrafficCfg); err != nil {
```

Comment:
Because we defined Encode function inside the provider package, so this `Decode` part should be there too.

Reply:
addressed on https://github.com/pipe-cd/pipe/pull/1462/commits/2d8762fa6c29d235761bf1bd50ad9657d5e088df 🙏